### PR TITLE
Make users choose role by name or prefix of name

### LINKED ISSUE #61: Make users choose role by name.
Roles should be displayed as their name instead of as a number for convenience.

### PR DESCRIPTION
**Pull request by Tanish2002**
Makes user able to select role with a Number, Role-Name or just the first letter of Role-Name
Preview: [here](https://i.fiery.me/z7yNt.png)
Fixes : #61

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -62,3 +62,4 @@ export DATABASE_URL=postgresql://user@localhost/trup
 - [kayew](https://github.com/kayew) for a more explicit setfetch message
 - [davidv171](https://github.com/davidv171) for mute command
 - [elkowar](https://github.com/elkowar) for blocklist command, media-logging in botlog and !poll multi
+- [Tanish2002](https://github.com/Tanish2002) for info command, cleaner !help and DM for ban and blocklist violation
```
Comment (buffet) [inline readme.md]: This is missing the role stuff itself.

File: command/role.go [hunk 2]
```diff
@@ -10,7 +10,7 @@ import (
 )
 
 const (
-	roleUsage = "role [number]"
+	roleUsage = "role [number/name/prefix of name]"
```
Comment (buffet) [inline command/role.go]: I'm not sure if mentioning the prefix helps or adds confusion here.

File: command/role.go [hunk 3]
```diff
@@ -31,16 +30,43 @@ func role(ctx *Context, args []string) {
...
-		ctx.Reply("Invalid number")
-		return
+	type colorRole struct {
+		ID   string `json:"id"`
+		Name string `json:"color"`
+	}
+	var colorRoles []colorRole
+	for i, _ := range ctx.Env.RoleColors {
+		color, _ := ctx.Session.State.Role(ctx.Message.GuildID, ctx.Env.RoleColors[i])
+		name := fmt.Sprintf("%s", (color.Name))
```
Comment (buffet) [inline command/role.go]: Wouldn't `name := color.Name` or something similar work here?

File: command/role.go [hunk 4]
```diff
@@ -31,16 +30,43 @@ func role(ctx *Context, args []string) {
...
+		r := colorRole{
+			ID:   color.ID,
+			Name: name,
+		}
+		colorRoles = append(colorRoles, r)
```
Comment (buffet) [inline command/role.go]: I think this initialization should be done during startup, not everytime the command is run.

File: command/role.go [hunk 5]
```diff
@@ -31,16 +30,43 @@ func role(ctx *Context, args []string) {
...
+
+	}
+	var roleID string
+	var addRole bool
+
+	for _, r := range colorRoles {
+		if args[1][0:1] == r.Name[0:1] || args[1] == r.Name {
```
Comment (buffet) [inline command/role.go]: Prefix vs first char. Maybe make that more clear, or add actual prefix support, not sure if that's an issue.

**Review by buffet — CHANGES_REQUESTED** [review]

File: readme.md (hunk 1 above)
Comment (Tanish2002) [reply readme.md]: ah yes dumb me

File: command/role.go (hunk 2 above)
Comment (Tanish2002) [reply command/role.go]: what else do you suggest?

File: command/role.go (hunk 4 above)
Comment (Tanish2002) [reply command/role.go]: got it

File: command/role.go (hunk 5 above)
Comment (Tanish2002) [reply command/role.go]: what do you mean? I don't get you

File: command/role.go (hunk 3 above)
Comment (Tanish2002) [reply command/role.go]: ah my bad I was actually using `%s\n` but for some reason I removed the new line but forgot to remove `Sprintf`

File: command/role.go (hunk 4 above)
Comment (Tanish2002) [reply command/role.go]: NVM, I don't know how I would implement this

Comment (davidv171) [reply command/role.go]: I don't think you get to just loop over RoleColors, I think we just store IDs in the environment. So for now, the environment variable either needs to be duplicated(to include colors and ids), or changed, or just left. 
@mlvzk thoughts?

Comment (mlvzk) [reply command/role.go]: You could intialize the color role structures in main.go's session ready handler - `discord.AddHandler(func(s *discordgo.Session, r *discordgo.Ready) {`, and put it in the Env struct(will require changing the type of RoleColors)

Comment (Tanish2002) [reply command/role.go]: okay

File: command/role.go [hunk 6]
```diff
@@ -31,25 +27,32 @@ func role(ctx *Context, args []string) {
...
+	if len(possibleRoles) == 0 {
+		ctx.Reply("Invalid role name")
 		return
-	} else if number < 0 || number >= len(ctx.Env.RoleColors) {
-		ctx.Reply("Invalid number")
+	} else if len(possibleRoles) == 1 {
+		roleID = possibleRoles[0]
+		addRole = true
+	} else if len(possibleRoles) > 1 {
+		ctx.Reply("Found more than 1 roles, try a more descriptive name")
```
Comment (buffet) [inline command/role.go]: I'd suggest something like
```
len == 0 {
    ...
    return
}
len > 1 {
    ...
    return
}

roleID = possibleRoles[0]
addRole = true ```
instead.

File: main.go [hunk 7]
```diff
@@ -432,3 +431,15 @@ func setStatus(s *discordgo.Session) {
 	update := discordgo.UpdateStatusData{Game: &game}
 	s.UpdateStatusComplex(update)
 }
+
+func rolesStuff(s *discordgo.Session, m *discordgo.MessageCreate) {
```
Comment (buffet) [inline main.go]: Is MessageCreate really the best thing to use here?

**Review by buffet — CHANGES_REQUESTED** [review]

File: command/role.go (hunk 6 above)
Comment (Tanish2002) [reply command/role.go]: okay

File: main.go (hunk 7 above)
Comment (Tanish2002) [reply main.go]: no

**Review by buffet — APPROVED** [review]

**Comment by buffet** [comment]
LGTM!

**Comment by eepykate** [comment]
wow october has a lot of free microwaves

File: main.go [hunk 8]
```diff
@@ -432,3 +431,19 @@ func setStatus(s *discordgo.Session) {
 	update := discordgo.UpdateStatusData{Game: &game}
 	s.UpdateStatusComplex(update)
 }
+
+func rolesStuff(s *discordgo.Session, g *discordgo.GuildCreate) {
```
Comment (mlvzk) [inline main.go]: This won't work, it needs to be
```suggestion
func rolesStuff(s *discordgo.Session, g *discordgo.Ready) {
```

also rename the function to initializeRoles

Comment (buffet) [reply main.go]: Ready doesn't work, as it gets called before the guilds are fetched.

GuildCreate gets called everytime the bot starts however.

Comment (Tanish2002) [reply main.go]: now it works :eye::lips::eye:

**Comment by buffet** [comment]
Can this be merged?

**Comment by mlvzk** [comment]
Thanks

**Pull request merged**